x-pack/plugins/security/server/lib/authorization/has_privileges.js

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { getClient } from '../../../../../server/lib/get_client_shield';
+import { DEFAULT_RESOURCE } from '../../../common/constants';
+import { getVersionPrivilege, getLoginPrivilege } from '../privileges';
+
+export const HAS_PRIVILEGES_RESULT = {
+  UNAUTHORIZED: Symbol(),
+  AUTHORIZED: Symbol(),
+  LEGACY: Symbol(),
+};
+
+export function hasPrivilegesWithServer(server) {
+  const callWithRequest = getClient(server).callWithRequest;
+
+  const config = server.config();
+  const kibanaVersion = config.get('pkg.version');
+  const application = config.get('xpack.security.rbac.application');
+  const kibanaIndex = config.get('kibana.index');
+
+  const loginPrivilege = getLoginPrivilege();
+  const versionPrivilege = getVersionPrivilege(kibanaVersion);
+
+  return function hasPrivilegesWithRequest(request) {
+
+    const hasApplicationPrivileges = async (privileges) => {
+      const privilegeCheck = await callWithRequest(request, 'shield.hasPrivileges', {
+        body: {
+          applications: [{
+            application,
+            resources: [DEFAULT_RESOURCE],
+            privileges
+          }]
+        }
+      });
+
+      const hasPrivileges = privilegeCheck.application[application][DEFAULT_RESOURCE];
```

Comment:
Given how critical this integration is with Elasticsearch, I think some response validation is in order on our end before we start trying to use the response. If Elasticsearch isn't giving us back exactly the shape of data we need, it would be good to fail abruptly as soon as we get the response rather than subtly later when we're using that value.

In other words, we should validate that things like `has_all_requested` exists and is a boolean, `application` is an object with the appropriate application id in it, etc. etc.

When it comes to security, the fewer instances of saying "well, that wasn't expected, so I don't know what the behavior would be off hand", the better.

The same suggestion is relevant for the legacy fallback, though there isn't a whole lot to validate there.

These scenarios should be tested too.